#!/usr/bin/env python

# Copyright 2018 National Technology & Engineering Solutions of Sandia, LLC
# (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the U.S.
# Government retains certain rights in this software.

#RUNTEST: trig

import sys
sys.dont_write_bytecode = True
sys.excepthook = sys.__excepthook__
import os
from os.path import abspath
import time
import unittest
from textwrap import dedent
import subprocess

try:
  from StringIO import StringIO
except Exception:
  from io import StringIO

import trigtestutils as trigutil
import testutils as util
from testutils import print3

import remotepython
from remotepython import RemotePython


rpy = sys.executable
#rpy='/home/rrdrake/bin/Linux/py266/bin/python'
#rpy='/home/rrdrake/bin/Linux/py365/bin/python'


class TestTimeoutError( Exception ):
    pass


class remote_process_tests( trigutil.trigTestCase ):

    def test_communicating_with_a_subprocess(self):
        ""
        util.writefile( 'repeater', """
            import sys 

            while True:
                next_line = sys.stdin.readline()
                if not next_line:
                    break
                sys.stdout.write( 'RCV: '+next_line )
                sys.stdout.flush()
            """ )
        time.sleep(1)

        proc = subprocess.Popen( sys.executable+' repeater', 
                                 shell=True,
                                 stdin=subprocess.PIPE,
                                 stdout=subprocess.PIPE )

        for i in range(10):
            proc.stdin.write( remotepython.make_bytes( '%d\n' % i ) )
            proc.stdin.flush()
            output = remotepython.make_string( proc.stdout.readline() )
            print ( 'OUT='+repr(output.rstrip()) )

        proc.terminate()

    def test_send_and_recv_data_with_a_subprocess(self):
        ""
        util.writefile( 'repeater', """
            import sys 

            while True:
                next_line = sys.stdin.readline()
                if not next_line:
                    break
                sys.stdout.write( 'OUT: '+next_line )
                sys.stdout.flush()
                sys.stderr.write( 'ERR: '+next_line )
                sys.stderr.flush()
            """ )
        time.sleep(1)

        rp = remotepython.RemoteProcess( [ sys.executable, abspath('repeater') ]) 

        assert rp.isAlive()

        inp = rp.getStdIn()
        out = rp.getStdOut()
        err = rp.getStdErr()

        outL = []
        errL = []
        for word in [ 'hello', 'world' ]:
            inp.write( remotepython.make_bytes( word+'\n' ) )
            inp.flush()
            outL.append( remotepython.make_string( out.readline() ) )
            errL.append( remotepython.make_string( err.readline() ) )

        assert outL == [ 'OUT: hello\n', 'OUT: world\n' ]
        assert errL == [ 'ERR: hello\n', 'ERR: world\n' ]

        assert rp.isAlive()
        rp.close()
        assert not rp.isAlive()


class command_writer_tests( trigutil.trigTestCase ):

    def write_to_file(self, logfp=None):
        ""
        with open( 'ofile', 'wb' ) as fp:
            rw = remotepython.ReprWriter( fp, logfp )
            rw.write( ' some\ndata ' )
            rw.write( 'more data\n' )

        time.sleep(1)

        lines = util.readfile( 'ofile' ).splitlines()
        assert len( lines ) == 2
        assert eval( lines[0] ) == ' some\ndata '
        assert eval( lines[1] ) == 'more data\n'

    def test_writer_sends_repr_strings_plus_newline(self):
        ""
        self.write_to_file()

    def test_writer_logging_to_file(self):
        ""
        with open( 'logfile', 'wt' ) as logfp:
            self.write_to_file( logfp )

        time.sleep(1)

        lines = util.readfile( 'logfile' ).splitlines()
        assert len( lines ) == 3
        assert lines[0] == 'SND:  some'
        assert lines[1] == '  2: data '
        assert lines[2] == 'SND: more data'


class async_reader_tests( trigutil.trigTestCase ):

    def make_pipe(self):
        ""
        fdr,fdw = os.pipe()

        if sys.version_info[0] < 3:
            readfp = os.fdopen( os.dup( fdr ), 'rb' )
        else:
            readfp = os.fdopen( fdr, 'rb', closefd=False )

        return fdr, fdw, readfp

    def test_reading_lines_from_a_file_descriptor(self):
        ""
        fdr,fdw,readfp = self.make_pipe()

        try:
            ar = remotepython.AsynchronousReader( readfp )
            write_to_fd( fdw, 'line of data\n' )
            write_to_fd( fdw, 'line two\n' )
            line1 = get_line_from_reader( ar )
            line2 = get_line_from_reader( ar )
        finally:
            force_close_fd( fdw )
            force_close_fd( fdr )
            # readfp.close()  # generates error messages with Python 2

        assert line1 == 'line of data\n'
        assert line2 == 'line two\n'

    def test_flushing_lines_from_reader(self):
        ""
        fdr,fdw,readfp = self.make_pipe()

        try:
            ar = remotepython.AsynchronousReader( readfp )
            write_to_fd( fdw, 'line of data\n' )
            write_to_fd( fdw, 'line two\n' )
            time.sleep(1)
            lines = ar.flushLines()
            ar = None
        finally:
            force_close_fd( fdw )
            force_close_fd( fdr )

        # print3( lines )
        assert lines.splitlines() == [ 'line of data', 'line two' ]

    def run_logfile(self, prefix=None):
        ""
        fdr,fdw,readfp = self.make_pipe()

        try:
            with open( 'logfile', 'wt' ) as logfp:
                if prefix:
                    ar = remotepython.AsynchronousReader( readfp, logfp, prefix )
                else:
                    ar = remotepython.AsynchronousReader( readfp, logfp )
                write_to_fd( fdw, 'line of data\n' )
                write_to_fd( fdw, 'line two\n' )
                time.sleep(1)
        finally:
            force_close_fd( fdw )
            force_close_fd( fdr )

    def test_logging_lines_to_a_file(self):
        ""
        self.run_logfile()
        time.sleep(1)

        log = util.readfile( 'logfile' )
        # print3( log )
        assert log.splitlines() == [ 'LOG: line of data', 'LOG: line two' ]

    def test_logging_lines_to_a_file_with_prefix(self):
        ""
        self.run_logfile( 'ABC' )
        time.sleep(1)

        log = util.readfile( 'logfile' )
        # print3( log )
        assert log.splitlines() == [ 'ABC: line of data', 'ABC: line two' ]


class bootstrap_functions( trigutil.trigTestCase ):

    def test_adding_lines_to_local_linecache(self):
        ""
        util.writefile( 'scr.py',
            remotepython.bootstrap_preamble + \
            dedent( """
                fn1 = _remotepython_add_eval_linecache( "line one" )
                fn2 = _remotepython_add_eval_linecache( "line two" )
                assert fn1 != fn2
                assert _remotepython_linecache[fn1] == [ "line one" ]
                assert _remotepython_linecache[fn2] == [ "line two" ]
                """ ) )
        time.sleep(1)

        x,out = util.runcmd( sys.executable+' '+abspath('scr.py') )
        # print3( out )

    def test_replacement_linecache_getline_function(self):
        ""
        util.writefile( 'mymod.py', """\
            import os, sys
            def myfunc():
                pass
            """ )
        util.writefile( 'scr.py',
            remotepython.bootstrap_preamble + \
            dedent( """
                import mymod

                fn1 = _remotepython_add_eval_linecache( "line one" )
                fn2 = _remotepython_add_eval_linecache( "line two" )
                assert linecache.getline( fn1, 1 ) == "line one"

                # if filename does not exist in cache or line number is
                # out of bounds somehow, return an empty string
                assert linecache.getline( 'somefakename', 5 ) == ""
                assert linecache.getline( fn1, 5 ) == ""

                assert linecache.getline( mymod.__file__, 2 ).strip() == "def myfunc():"
                """ ) )
        time.sleep(1)

        x,out = util.runcmd( sys.executable+' '+abspath('scr.py') )
        # print3( out )

    def test_add_a_module_from_a_string(self):
        ""
        mymod = dedent( """
            import sys
            def myfunc():
                return sys.version_info
            """ )
        util.writefile( 'scr.py',
            remotepython.bootstrap_preamble + \
            dedent( """
                _remotepython_add_module( 'mymod', """+repr(mymod)+""" )
                import sys
                import mymod
                assert mymod.myfunc() == sys.version_info
                """ ) )
        time.sleep(1)

        x,out = util.runcmd( sys.executable+' '+abspath('scr.py') )
        # print3( out )

    def test_executing_python_code(self):
        ""
        code = dedent( """
            import sys, os
            newvar = 'add this to globals'
            myvar = 24
            def myfunc():
                global myvar
                myvar = 42
            sys.stdout.write( 'AVALUE='+str(myvar)+os.linesep )
            myfunc()
            sys.stdout.write( 'BVALUE='+str(myvar)+os.linesep )
            sys.stdout.flush()
            """ )
        util.writefile( 'scr.py',
            remotepython.bootstrap_preamble + \
            dedent( """
                _remotepython_eval_lines( """+repr(code)+""" )
                print ( 'NEWVAR='+repr(newvar) )
                """ ) )
        time.sleep(1)

        x,out = util.runcmd( sys.executable+' '+abspath('scr.py') )
        # print3( out )
        assert 'AVALUE=24' in out and 'BVALUE=42' in out
        assert "NEWVAR='add this to globals'" in out

    def test_the_traceback_module_will_use_a_replacement_getline(self):
        """
        The traceback module uses the linecache module to get line information
        for printing tracebacks (such as for exceptions). Thus, the getline()
        function of linecache can be replaced to serve line info for code
        compiled from a string.

        Note that this does not work with an uncaught exception.
        """
        util.writefile( 'extmod.py', """\
            import os, sys
            def afunc():
                raise Exception( "fake exception" )
            """ )
        mymod = dedent( """
            import extmod
            def myfunc():
                extmod.afunc()
            """ )
        mycode = dedent( """
            def entryfunc():
                mymod.myfunc()
            """ )
        util.writefile( 'scr.py',
            remotepython.bootstrap_preamble + \
            dedent( """
                _remotepython_add_module( 'mymod', """+repr(mymod)+""" )

                import mymod

                fn = _remotepython_add_eval_linecache( """+repr(mycode)+""" )
                eval( compile( """+repr(mycode)+""", fn, "exec" ) )

                try:
                    entryfunc()
                except Exception:
                    traceback.print_exc()
                """ ) )
        time.sleep(1)

        x,out = util.runcmd( sys.executable+' '+abspath('scr.py') )
        # print3( out )

        excL = extract_exception_from_output( out )

        assert excL[0] == 'Traceback (most recent call last):'
        assert excL[1].startswith('  File ') and '/scr.py' in excL[1]
        assert excL[2] == '    entryfunc()'
        assert excL[3] == '  File "<remotecode1>", line 3, in entryfunc'
        assert excL[4] == '    mymod.myfunc()'
        assert excL[5] == '  File "<remotemodule_mymod>", line 4, in myfunc'
        assert excL[6] == '    extmod.afunc()'
        assert excL[7].startswith('  File ') and \
               excL[7].endswith('extmod.py", line 3, in afunc')
        assert excL[8] == '    raise Exception( "fake exception" )'
        assert excL[9] == 'Exception: fake exception'


class integration_tests( unittest.TestCase ):

    def setUp(self):
        util.setup_test()

    def test_connect_and_print_hello(self):
        ""
        rem = RemotePython( pythonexe=rpy )
        try:
            ok = rem.start(5)
            assert ok
            rem.execute( 'import os, sys',
                         'print ( "hello stdout" )',
                         'sys.stderr.write( "hello stderr"+os.linesep )',
                         'sys.stderr.flush()' )
            out1,err1 = get_line_from_remote( rem )
            out2,err2 = get_line_from_remote( rem )
        finally:
            rem.close()

        assert out1 != out2 and err1 != err2
        assert out1 == 'hello stdout\n' or out2 == 'hello stdout\n'
        assert err1 == 'hello stderr\n' or err2 == 'hello stderr\n'

    def test_timing_out_waiting_on_a_command(self):
        ""
        rem = RemotePython( pythonexe=rpy )
        try:
            rem.start()
            rem.execute( 'import time',
                         'time.sleep(10)' )
            get_line_from_remote( rem, timeout=5 )
        except TestTimeoutError:
            rem.close()
        else:
            rem.close()
            raise Exception( 'expected a timeout' )

    def test_check_if_subprocess_is_alive(self):
        ""
        rem = RemotePython( pythonexe=rpy )
        try:
            rem.start()
            chk1 = rem.isAlive()
            rem.execute( 'import time',
                         'time.sleep(2)',
                         'print ( "foobar" )' )
            chk2 = rem.isAlive()
            out1,err1 = get_line_from_remote( rem )
            chk3 = rem.isAlive()
            rem.execute( 'raise Exception("fake failure")' )
            time.sleep(1)
            chk4 = rem.isAlive()
            out2,err2 = flush_lines_from_remote( rem )
        finally:
            rem.close()
        chk5 = rem.isAlive()

        assert chk1 == True
        assert chk2 == True
        assert chk3 == True
        assert chk4 == False
        assert chk5 == False
        assert out1.strip() == 'foobar' and not err1
        assert not out2 and 'fake failure' in err2

    def test_connecting_with_ssh_but_using_a_fake_ssh_script(self):
        ""
        fakessh = write_fake_ssh_script()
        time.sleep(1)

        cmdL = remotepython.bootstrap_command( '/opt/super py',
                                               None,
                                               'foobar' )
        assert 'foobar' not in ''.join( cmdL )
        assert cmdL[0] == '/opt/super py'

        cmdL = remotepython.bootstrap_command( '/opt/super py',
                                               'sparky',
                                               '/opt/foobar -o baz' )
        assert cmdL[:4] == [ '/opt/foobar', '-o', 'baz', 'sparky' ]
        assert '/opt/super py' in ''.join( cmdL )

        rem = RemotePython( 'sparky', pythonexe=rpy, sshcmd=fakessh )
        try:
            rem.start( timeout=5 )
            rem.execute( 'print ("hello")' )
            out,err = get_line_from_remote( rem )
        finally:
            rem.close()
        assert out.strip() == 'hello' and not err

    def test_that_output_during_connection_is_available(self):
        ""
        fakessh = write_fake_ssh_script( to_stdout='outval1\\n outval2',
                                         to_stderr='errval1\\n errval2' )
        time.sleep(1)

        rem = RemotePython( 'sparky', pythonexe=rpy, sshcmd=fakessh )
        try:
            rem.start()
            rem.execute( 'print ("hello")' )
            out,err = get_line_from_remote( rem )
            conn = rem.getStartupOutput()
        finally:
            rem.close()
        assert out.strip() == 'hello'
        assert not err
        assert 'outval1' in conn and 'outval2' in conn
        assert 'errval1' in conn and 'errval2' in conn
        assert len( util.greplines( 'out*', conn ) ) == 2
        assert len( util.greplines( 'err*', conn ) ) == 2

    def test_when_ssh_immediately_crashes(self):
        ""
        fakessh = write_fake_ssh_script( die_fast=True,
                                         to_stdout='junkout',
                                         to_stderr='junkerr' )
        time.sleep(1)

        rem = RemotePython( 'sparky', pythonexe=rpy, sshcmd=fakessh )
        ok = rem.start()
        assert not ok
        conn = rem.getStartupOutput()
        assert conn and "connection failed" in conn
        assert "fake crash" in conn
        assert "junkout" in conn
        assert "junkerr" in conn
        assert not rem.isAlive()

    def test_when_ssh_connection_times_out(self):
        ""
        fakessh = write_fake_ssh_script( fake_pause=10 )
        time.sleep(1)

        rem = RemotePython( 'sparky', pythonexe=rpy, sshcmd=fakessh )
        ok = rem.start( timeout=5.5 )
        assert not ok
        conn = rem.getStartupOutput()
        assert conn and "timed out" in conn
        assert "5.5" in conn
        assert not rem.isAlive()

    def test_logging_all_traffic(self):
        ""
        fakessh = write_fake_ssh_script( to_stdout='connect output data',
                                         to_stderr='connect error data' )
        time.sleep(1)

        fp = open( 'logfile', 'wt' )
        rem = RemotePython( 'sparky',
                            pythonexe=rpy,
                            sshcmd=fakessh,
                            logfile=fp )
        try:
            rem.start()
            rem.execute( 'print ( "hello 10-4" )',
                         'import os, sys',
                         'sys.stderr.write( "some error output 357"+os.linesep )',
                         'sys.stderr.flush()',
                         'raise Exception( "fake exception" )' )
            while rem.isAlive():
                time.sleep(0.5)
        finally:
            rem.close()
            fp.close()

        assert len( util.grepfiles( 'CMD:*fakessh', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'SND:*RemotePython', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'OUT: connect output data', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'ERR: connect error data', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'OUT: RemotePython', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'SND: *hello 10-4', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'OUT: hello 10-4', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'ERR: some error output 357', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'ERR: Traceback', 'logfile' ) ) > 0
        assert len( util.grepfiles( 'ERR: Exception*fake exception', 'logfile' ) ) == 1

    def test_giving_a_filename_for_the_logfile(self):
        ""
        fakessh = write_fake_ssh_script()
        time.sleep(1)

        rem = RemotePython( 'sparky',
                            pythonexe=rpy,
                            sshcmd=fakessh,
                            logfile='logfile' )
        try:
            rem.start()
            rem.execute( 'print ( "hello good buddy" )' )
            rem.execute( 'sys.exit(0)' )
            while rem.isAlive():
                time.sleep(0.5)
        finally:
            rem.close()

        assert len( util.grepfiles( 'SND: *hello good buddy', 'logfile' ) ) == 1
        assert len( util.grepfiles( 'OUT: hello good buddy', 'logfile' ) ) == 1

    def test_that_remote_traceback_includes_code_lines(self):
        ""
        fakessh = write_fake_ssh_script()
        time.sleep(1)

        fp = open( 'logfile', 'wt' )
        rem = RemotePython( 'sparky',
                            pythonexe=rpy,
                            sshcmd=fakessh,
                            logfile=fp )
        try:
            rem.start()
            rem.execute( 'import os, sys',
                         'def func1(arg):',
                         '    func2(arg)',
                         'def func2(arg):',
                         '    raise Exception( "fake exception "+arg )' )
            rem.execute( 'if True:',
                         '    func1( "foobar" )' )
            while rem.isAlive():
                time.sleep(0.5)
        finally:
            rem.close()
            fp.close()

        # Traceback (most recent call last):
        #   File "<bootstrap>", line 26, in <module>
        #   File "<remotecode3>", line 2, in <module>
        #     func1( "foobar" )
        #   File "<remotecode2>", line 3, in func1
        #     func2(arg)
        #   File "<remotecode2>", line 5, in func2
        #     raise Exception( "fake exception "+arg )
        # Exception: fake exception foobar
        exc0 = extract_exception_from_logfile(0)
        # print3( exc0 )
        assert len( util.greplines( 'File ', exc0 ) ) == 4
        assert 'func1( "foobar" )' in exc0
        assert 'func2(arg)' in exc0
        assert len( util.greplines( 'File *line 5, in func2', exc0 ) ) == 1
        assert 'raise Exception( "fake exception "+arg )' in exc0
        assert 'fake exception foobar' in exc0.splitlines()[-1]

        # Traceback (most recent call last):
        #   File "/scratch/rrdrake/temp/tmpdir_remotepython_tests/fakessh", line 7, in <module>
        #     subprocess.check_call( cmdL )
        #   File "/usr/lib64/python2.6/subprocess.py", line 505, in check_call
        #     raise CalledProcessError(retcode, cmd)
        # subprocess.CalledProcessError: Command '['/usr/bin/python', '-u', '-E', '-c', 'import sys; eval( compile( eval( sys.stdin.readline() ), "<bootstrap>", "exec" ) )']' returned non-zero exit status 1
        exc1 = extract_exception_from_logfile(1)
        # print3( exc1 )
        assert len( util.greplines( 'File ', exc0 ) ) > 0
        assert 'returned non-zero exit status' in exc1.splitlines()[-1]

    def test_adding_a_module_then_importing_it(self):
        ""
        util.writefile( 'subdir/afile.txt', 'some contents' )
        time.sleep(1)

        cwd = os.getcwd()

        rem = RemotePython( pythonexe=rpy )
        try:
            ok = rem.start()
            assert ok
            rem.addModule( 'mymod',
                    'import os, sys',
                    'def myfunc( arg ):',
                    '    os.chdir( "'+abspath('subdir')+'" )',
                    '    print ( "LISTDIR="+repr(os.listdir(".")) )' )
            rem.execute( 'import mymod' )
            rem.execute( 'mymod.myfunc(42)' )
            out,err = get_line_from_remote( rem )
        finally:
            rem.close()
        assert err == None
        fL = eval( out.strip().split( 'LISTDIR=', 1 )[1] )
        assert fL == ['afile.txt']

        assert os.path.samefile( cwd, os.getcwd() )

    def test_remote_module_traceback_should_have_code_lines(self):
        ""
        rem = RemotePython( pythonexe=rpy, logfile='logfile' )
        try:
            ok = rem.start()
            assert ok
            rem.addModule( 'mymod1',
                    'import os, sys',
                    'def myfunc( arg ):',
                    '    if arg:',
                    '        raise Exception( "fake exception" )' )
            rem.addModule( 'mymod2',
                    'import os',
                    'import sys',
                    'import mymod1',
                    'def myfunc():',
                    '    mymod1.myfunc( 42 )' )
            rem.execute( 'import mymod2',
                         'mymod2.myfunc()' )
            while rem.isAlive():
                time.sleep(0.5)
        finally:
            rem.close()

        # Traceback (most recent call last):
        #   File "<bootstrap>", line 42, in <module>
        #   File "<remotecode4>", line 2, in <module>
        #     mymod2.myfunc()
        #   File "<remotemodule_mymod2>", line 5, in myfunc
        #     mymod1.myfunc( 42 )
        #   File "<remotemodule_mymod1>", line 4, in myfunc
        #     raise Exception( "fake exception" )
        # Exception: fake exception
        exc = extract_exception_from_logfile(0)
        # print3( exc )
        assert len( util.greplines( 'File ', exc ) ) == 4
        assert 'mymod2.myfunc()' in exc
        assert 'mymod1.myfunc( 42 )' in exc
        assert 'raise Exception( "fake exception" )' in exc
        assert len( util.greplines( 'File *mymod2*line 5, in myfunc', exc ) ) == 1
        assert 'fake exception' in exc.splitlines()[-1]

    def test_adding_a_module_with_syntax_error_shows_code_lines(self):
        ""
        rem = RemotePython( pythonexe=rpy, logfile='logfile' )
        try:
            ok = rem.start()
            assert ok
            rem.addModule( 'mymod',
                    'import os, sys',
                    'if 1 = 2:',
                    '    pass' )
            while rem.isAlive():
                time.sleep(0.5)
            out,err = flush_lines_from_remote( rem )
        finally:
            rem.close()

        assert not out
        print3( err )
        accum = []
        for line in err.splitlines():
            line = line.strip()
            if 'remotemodule_mymod' in line and 'line 2' in line:
                accum.append( 'loc' )
            elif line.startswith( 'if 1 = 2:' ):
                accum.append( 'if' )
            elif line.startswith( '^' ):
                accum.append( 'carrot' )
            elif line.startswith( 'SyntaxError: invalid syntax' ):
                accum.append( 'err' )
        assert accum == ['loc','if','carrot','err']

    def test_using_real_ssh(self):
        ""
        return  # magic: make execution of "real ssh" tests an option

        rem = RemotePython( os.uname()[1], pythonexe=rpy )
        #rem = RemotePython( 'chama' )
        try:
            ok = rem.start( timeout=5 )
            if ok:
                rem.execute( 'print ( "hello world" )' )
                out,err = get_line_from_remote( rem )
        finally:
            rem.close()
        print3( 'startup output\n', rem.getStartupOutput() )
        assert ok
        assert out.strip() == 'hello world'
        assert err == None

    def test_stressing_get_lines_and_many_sequential_connections(self):
        ""
        return  # magic: make execution of performance testing an option

        count = 10000
        for i in range(count):
            rem = RemotePython( pythonexe=rpy )
            try:
                rem.start()
                rem.execute( 'import time',
                             'print ( "hello" )',
                             '#time.sleep(0.3)',
                             'print ( "world" )' )
                out = rem.getOutputLine()
                while out == None or out.strip() != 'world':
                    out = rem.getOutputLine()
            finally:
                rem.close()
            print ( 'i='+str(i) )


def write_to_fd( fd, data ):
    ""
    if sys.version_info[0] < 3:
        os.write( fd, data )
    else:
        os.write( fd, data.encode() )


def force_close_fd( fd ):
    ""
    try:
        os.close( fd )
    except Exception:
        pass


def list_fds():
    """
    for linux OS, this can be used to debug file descriptor leaks
    """
    fddir = '/proc/self/fd'
    for fn in os.listdir(fddir):
        pn = os.path.join( fddir, fn )
        if os.path.islink(pn):
            print3( pn, '->', os.readlink(pn) )
        else:
            print3( pn )


def extract_exception_from_logfile( exc_index=0 ):
    ""
    cnt = -1
    exc = ''
    with open( 'logfile', 'rt' ) as fp:
        for line in fp:
            if cnt == exc_index:
                exc += line.split( 'ERR: ', 1 )[1]
                if not line.startswith( 'ERR:   ' ):
                    break
            elif line.startswith( 'ERR: Traceback' ):
                cnt += 1
                if cnt == exc_index:
                    exc += line.split( 'ERR: ', 1 )[1]

    return exc


def extract_exception_from_output( output, exc_index=0 ):
    ""
    cnt = -1
    exc = None
    for line in output.splitlines():
        if not exc:
            if line.startswith( 'Traceback ' ):
                exc = [ line.rstrip() ]
                cnt += 1
        elif line.startswith( 'Exception: ' ):
            exc.append( line.rstrip() )
            if cnt == exc_index:
                break
            exc = None
        else:
            exc.append( line.rstrip() )

    return exc


def get_line_from_reader( rdr, timeout=5 ):
    ""
    tstart = time.time()

    while True:

        line = rdr.getLine()

        if line != None:
            break
        elif time.time()-tstart > timeout:
            raise TestTimeoutError( 'timeout waiting for output' )
        else:
            time.sleep( 0.5 )

    return line


def get_line_from_remote( rem, timeout=5 ):
    ""
    tstart = time.time()
    while True:

        line = rem.getOutputLine()
        if line != None:
            return line, None

        line = rem.getErrorLine()
        if line != None:
            return None, line

        if time.time()-tstart > timeout:
            raise TestTimeoutError( 'timeout waiting for output' )

        time.sleep(0.5)


def flush_lines_from_remote( rem ):
    ""
    outbuf = None
    errbuf = None

    while True:
        out = rem.getOutputLine()
        err = rem.getErrorLine()
        if out:
            if outbuf == None:
                outbuf = out
            else:
                outbuf += out
        if err:
            if errbuf == None:
                errbuf = err
            else:
                errbuf += err

        if not out and not err:
            break

    return outbuf, errbuf


def write_fake_ssh_script( die_fast=False, fake_pause=None,
                           to_stdout=None, to_stderr=None ):
    ""
    lineL = [ '#!'+sys.executable,
              'import sys, os, subprocess, shlex, time' ]

    if to_stdout:
        lineL.append( 'sys.stdout.write( "'+to_stdout+'"+os.linesep )' )
        lineL.append( 'sys.stdout.flush()' )

    if to_stderr:
        lineL.append( 'sys.stderr.write( "'+to_stderr+'"+os.linesep )' )
        lineL.append( 'sys.stderr.flush()' )

    if die_fast:
        lineL.append( 'raise Exception( "fake crash" )' )

    if fake_pause and fake_pause > 0:
        lineL.append( 'time.sleep( '+str(fake_pause)+' )' )

    lineL.extend( [ 'cmdL = []',
                    'for arg in sys.argv[2:]:  # exclude the machine name',
                    '    cmdL.extend( shlex.split( arg ) )',
                    'subprocess.check_call( cmdL )' ] )

    util.writescript( 'fakessh', '\n'.join( lineL ) )

    return abspath( 'fakessh' )


###########################################################################

util.run_test_cases( sys.argv, sys.modules[__name__] )
